tkg/client/upgrade_cluster.go

```diff
@@ -181,18 +181,35 @@ func (c *TkgClient) UpgradeCluster(options *UpgradeClusterOptions) error {
 	return c.DoLegacyClusterUpgrade(regionalClusterClient, currentClusterClient, options)
 }
 
-func (c *TkgClient) DoLegacyClusterUpgrade(regionalClusterClient, currentClusterClient clusterclient.Client, options *UpgradeClusterOptions) error {
+func (c *TkgClient) DoLegacyClusterUpgrade(regionalClusterClient, currentClusterClient clusterclient.Client, options *UpgradeClusterOptions) (err error) {
 	log.Info("Verifying kubernetes version...")
-	err := c.verifyK8sVersion(currentClusterClient, options.KubernetesVersion)
+	err = c.verifyK8sVersion(currentClusterClient, options.KubernetesVersion)
 	if err != nil {
 		return errors.Wrap(err, "kubernetes version verification failed")
 	}
 
+	// backup tkr version for current cluster
+	currentClusterInfo := regionalClusterClient.GetClusterStatusInfo(options.ClusterName, options.Namespace, currentClusterClient)
+	currentTkrVersion := currentClusterInfo.ClusterObject.Labels["tanzuKubernetesRelease"]
+
+	log.Infof("Upgrading kubernetes cluster from `%v` to `%v` version", currentTkrVersion, options.TkrVersion)
+
 	err = c.addKubernetesReleaseLabel(regionalClusterClient, options)
 	if err != nil {
 		return errors.Wrapf(err, "unable to patch the cluster object with TanzuKubernetesRelease label")
 	}
 
+	// restore tkr version if upgrade failed after tkr label updated
+	defer func() {
+		if err != nil {
+			log.Infof("Restoring kubernetes cluster to tkr `%v` version", currentTkrVersion)
+			options.TkrVersion = currentTkrVersion
```

Comment:
besides @12345lcr 's comments, curious what will be the suggested actions for users after setting tkr version back to the original version? upgrade again directly or do some "repair" then upgrade? if user need to do some manual "repair", seems set tkr version back could be part of that as well

Reply:
> upgrade again directly or do some "repair" then upgrade?
either way should be good, depends on what kind of error they meet. But if the version is not set back, customer will not be able to run upgrade again since the current version is overwritten and tanzu cli does not support upgrade to "old" versions.